__builtin_constant_p: properly test, fix behaviour

tautschnig opened the pull request:
We previously only compile-time tested what required a run-time check of assertions, so move this to the "cbmc" regression test suite. Also, extend it by behaviour that distinguishes it from actual C-standard specified constant expressions. Finally, fix the behaviour for string literals.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Mimic GCC/Clang simplification behaviour when type checking ?:

Neither GCC nor Clang simplify expressions involving symbols even when
the result would always be 0, which our simplifier figure out.

Fixes: #7927

kroening commented:
We will eventually need explicit constant folding logic in the C frontend, as opposed to relying on the simplifier. I expect that the latter will always be able to simplify more.

kroening commented:
This PR needs more contemplation. The PR changes the typing, but would appear to be about constant folding.
Maybe the presence of symbols doesn't change the typing, but does affect constant folding.

tautschnig committed:
Move is_compile_time_constantt to top of file

This is useful in various methods and should be available to all of
them.

tautschnig committed:
Move is_compile_time_constantt to top of file

This is useful in various methods and should be available to all of
them.

tautschnig commented:
> This PR needs more contemplation. The PR changes the typing, but would appear to be about constant folding. Maybe the presence of symbols doesn't change the typing, but does affect constant folding.

@kroening: Please take another look, I hope to have adjusted this in line with your feedback as it now uses (compile-time) constant folding rules.

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -44,6 +44,94 @@ Author: Daniel Kroening, kroening@kroening.com
+/// Architecturally similar to \ref can_forward_propagatet, but specialized for
+/// what is a constexpr, i.e., an expression that can be fully evaluated at
+/// compile time.
+class is_compile_time_constantt
```
May I suggest explicitly referencing the standard, and where the logic deviates from the standard to mimic gcc or clang.

tautschnig replied:
This is now done.

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1601,24 +1689,31 @@ void c_typecheck_baset::typecheck_expr_trinary(if_exprt &expr)
+  auto compile_time_null_pointer = [](const exprt &e, const namespacet &ns) {
+    if(!is_compile_time_constantt(ns)(e))
+      return false;
+    auto s = simplify_expr(e, ns);
```
I'd be worried about the use of the simplifier: This may prevent us in the future from making the simplifier smarter, as that might then make additional expressions constant that aren't meant to be.

tautschnig replied:
While I agree that we have this problem at the moment (i.e., without the changes proposed in this PR), doesn't the use of `is_compile_time-constantt` before invoking the simplifier address this risk?

kroening replied:
Would we dare to assert that `s` is a constant?

tautschnig replied:
No: `s` could be address-of (which is a permissible constant expression under C11 6.6).

kroening replied:
Can we assert `is_compile_time_constantt(ns)(s)`?

tautschnig replied:
Added!

tautschnig committed:
Expand documentation of is_compile_time_constantt

C11 Section 6.6 "Constant expressions" defines what are required to be
constant expressions, and permits implementations to pick additional
constant expressions.

tautschnig committed:
Expand documentation of is_compile_time_constantt

C11 Section 6.6 "Constant expressions" defines what are required to be
constant expressions, and permits implementations to pick additional
constant expressions.

kroening commented on regression/ansi-c/sizeof6/main.c:
```diff
@@ -0,0 +1,23 @@
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(i ? 0ll : 0ll)) : (int *)1)), "");
+  _Static_assert(
+    sizeof(int) != sizeof(*(1 ? ((void *)(0 ? i : 0ll)) : (int *)1)), "");
```
These appear to assert that some part of the subexpression is a compile time constant. May I suggest to use `__builtin_constant_p` for that. The sizeof assertion is a distraction.

tautschnig replied:
This, I'm afraid, just uncovered another challenge: the above "true" cases aren't deemed null-pointer constants by the compilers (and, therefore, the type of the `?:` is `void*` rather than `int*`). `__builtin_constant_p`, however, does return `1` for the very same expressions!

tautschnig replied:
I have added a commit that better exercises (and extends) an existing test to also document (and test) these differences.

kroening replied:
Uhhh, sad

kroening replied:
Might be worth filing a bug report with those compilers.

tautschnig committed:
__builtin_constant_p: properly test, fix behaviour

We previously only compile-time tested what required a run-time check of
assertions, so move this to the "cbmc" regression test suite. Also,
extend it by behaviour that distinguishes it from actual C-standard
specified constant expressions. Finally, fix the behaviour for string
literals.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
__builtin_constant_p: properly test, fix behaviour

We previously only compile-time tested what required a run-time check of
assertions, so move this to the "cbmc" regression test suite. Also,
extend it by behaviour that distinguishes it from actual C-standard
specified constant expressions. Finally, fix the behaviour for string
literals.

The set of expressions Clang constant-folds for __builtin_constant_p
omitted subtraction, unary minus, division, remainder, and the shift
operators. Per C11 6.6 these are all permitted in an integer constant
expression, and both GCC and Clang fold them, but the Clang
__builtin_constant_p path applies clang_is_constant_foldedt to the
unsimplified argument, so without these operators it wrongly reported
e.g. __builtin_constant_p(1 - 1) and __builtin_constant_p(0 << 3) as
non-constant in Clang mode. Add them to clang_is_constant_foldedt only,
keeping them out of the shared is_compile_time_constantt base predicate:
make_constant consults that base, and adding unfoldable operators there
would make it accept expressions the simplifier cannot fold (e.g. a
file-scope `int a[1/0];`), which it must continue to reject. So
make_constant's behaviour is unchanged for those operators.

A string literal, by contrast, is a legitimate address constant per C11
6.6, so ID_string_constant is intentionally accepted by the shared base
predicate (and hence by make_constant), not confined to
clang_is_constant_foldedt; this is what fixes __builtin_constant_p of a
string literal in Clang mode, where the argument arrives as a bare
ID_string_constant.  Non-integer contexts (e.g. `int a["abc"];`) remain
rejected by the index-type conversion.

is_compile_time_constantt is now a polymorphic base (clang_is_constant_foldedt
overrides is_constant); give it a virtual destructor. The condition
simplified in the Clang ternary and &&/|| short-circuit handling is not
guaranteed to be a constant_exprt (a string literal or infinity is
accepted by is_constant but not folded), so fall back to "not constant"
rather than asserting; the shared bool-vs-zero truthiness check is
factored into is_constant_true().

regression/ansi-c/builtin_constant_p_folding exercises the Clang
constant-folding path via the ansi-c-clang profile (the cbmc suite runs
in GCC mode and would not catch a Clang-only regression), including the
ternary and &&/|| short-circuit branches and the address-of rule.
regression/ansi-c/make_constant_div_by_zero pins down that make_constant
still rejects an unfoldable `1/0` as a file-scope array size.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening commented:
Apologies that the scope of this PR is getting pushed larger, by me. Does it make sense to replace the minimal case split done for `__builtin_constant_p` by an invocation of `is_compile_time_constantt(ns)` plus some special cases for when they differ?

tautschnig commented:
> Apologies that the scope of this PR is getting pushed larger, by me. Does it make sense to replace the minimal case split done for `__builtin_constant_p` by an invocation of `is_compile_time_constantt(ns)` plus some special cases for when they differ?

I'll take care of this!

tautschnig commented:
> > Apologies that the scope of this PR is getting pushed larger, by me. Does it make sense to replace the minimal case split done for `__builtin_constant_p` by an invocation of `is_compile_time_constantt(ns)` plus some special cases for when they differ?
> 
> I'll take care of this!

I believe this is now done.

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3594,27 +3725,35 @@ exprt c_typecheck_baset::do_special_functions(
     // do not typecheck the argument - it is never evaluated, and thus side
```
The comment would appear to be from the time when the type checker has taken care of side effects. The argument is type checked.

tautschnig replied:
Removed.

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3594,27 +3725,35 @@ exprt c_typecheck_baset::do_special_functions(
     else
-      is_constant=tmp1.is_constant();
+    {
+      // try to produce constant
```
`is_compile_time_constantt` is too far off in the GCC/MSVC case?

tautschnig replied:
`__builtin_constant_p` doesn't exist with MSVC, so it'd be GCC only. For that, however, the simplifier appears to be much closer: GCC happily takes `x - x`, `x * 0` as constants. But then we could still do what I now (in my latest push to this PR) ended up doing: produce a variant of `is_compile_time_constantt` for a specific compiler (`clang_is_constant_foldedt`).

tautschnig commented:
I have now moved the original starting point of this PR (fixing behaviour of `?:` to align with what GCC does) to #9052. My rationale is that most of the discussion on this PR wasn't as much about that initial fix, but rather the follow-up work for `__builtin_constant_p`.
